### PR TITLE
Add databaseIndex implementation in`storage`

### PR DESCRIPTION
**Pull request by prateek**
- Adds `databaseIndex` and `databaseIndexInsertQueue` construct and implementation. 

NB: In an effort to keep reviews small, i'm using `FOLLOWUP(prateek):...` littered in the code to indicate changes that will follow in subsequent PRs.

/cc @robskillington @richardartoul @jeromefroe @xichen2020

### CODE REVIEW

File: storage/index.go [hunk 1]
```diff
@@ -21,61 +21,260 @@
...
+// nolint: deadcode
+func newDatabaseIndex(
+	fn newDatabaseIndexInsertQueueFn,
+	opts index.Options,
+) (databaseIndex, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	seg, err := mem.New(1, opts.MemSegmentOptions())
```
Comment (richardartoul) [inline storage/index.go]: SUPER nit, but seems like 1 is a unique identifier for this particular index? I would say maybe make it a constant, but if we're moving to namespace-level indices...

Whats the id for?

File: storage/index.go [hunk 2]
```diff
@@ -21,61 +21,260 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	idx := &dbIndex{
+		insertMode: dbIndexInsertAsync,
```
Comment (richardartoul) [inline storage/index.go]: might be nice to make this wireable from the YAML config

File: storage/index.go [hunk 3]
```diff
@@ -21,61 +21,260 @@
...
+}
+
+func (i *dbIndex) writeBatch(inserts []dbIndexInsert) error {
+	// NB(prateek): we use a read lock to guard against mutation of the
+	// dbIndexBlock currently active, mutations within the underlying
+	// dbIndexBlock are guarded by primitives internal to it.
+	i.RLock()
+	defer i.RUnlock()
+
+	if !i.isOpenWithRLock() {
```
Comment (richardartoul) [inline storage/index.go]: Is skipping the OnFinalize calls really necessary? Seems like a potential source of shutdown-bugs depending on what people are putting in the OnIndexFinalize method

File: storage/index.go [hunk 4]
```diff
@@ -21,61 +21,260 @@
...
+	opts        index.Options
 }
 
-func newDatabaseIndex(o Options) (databaseIndex, error) {
-	return &dbIndex{
-		opts: o,
-	}, nil
+// nolint: deadcode
+func newDatabaseIndex(
+	fn newDatabaseIndexInsertQueueFn,
```
Comment (richardartoul) [inline storage/index.go]: A slightly more verbose name here would be nice, a spent a little time trying to track down what fn was below. Couldn't figure it out until I opened your branch in my editor

File: storage/index.go [hunk 5]
```diff
@@ -21,61 +21,260 @@
...
+
+func (i *dbIndex) writeBatch(inserts []dbIndexInsert) error {
+	// NB(prateek): we use a read lock to guard against mutation of the
+	// dbIndexBlock currently active, mutations within the underlying
+	// dbIndexBlock are guarded by primitives internal to it.
+	i.RLock()
+	defer i.RUnlock()
+
+	if !i.isOpenWithRLock() {
+		// NB(prateek): skip calling any of the `OnIndexFinalize` methods
```
Comment (richardartoul) [inline storage/index.go]: Did this make a noticeable difference in performance at any point? I think I understand what you're trying to do, but I spent enough time trying to decipher this comment that I wonder if its even worth doing this.

File: storage/index.go [hunk 6]
```diff
@@ -21,61 +21,260 @@
...
+	}
+
+	if numErr != 0 {
+		i.metrics.AsyncInsertErrors.Inc(numErr)
+	}
+
+	return nil
 }
 
 func (i *dbIndex) Write(
```
Comment (richardartoul) [inline storage/index.go]: The "normal path" will not use this method right? It'll go shard insert queue --> indexQueue --> writeBatch right?

File: storage/index.go [hunk 7]
```diff
@@ -21,61 +21,260 @@
...
+	}
 
-func (n dbIndexNoOp) Write(context.Context, ident.ID, ident.ID, ident.TagIterator) error {
-	return nil
+	i.state = dbIndexStateClosed
+	return i.insertQueue.Stop()
+}
+
+func (i *dbIndex) doc(ns, id ident.ID, tags ident.Tags) (doc.Document, error) {
+	fields := make([]doc.Field, 0, 1+len(tags))
```
Comment (richardartoul) [inline storage/index.go]: Any opportunity for pooling this, or does the index own it forever?

File: storage/index.go [hunk 8]
```diff
@@ -21,61 +21,260 @@
...
+	fields = append(fields, doc.Field{
+		Name:      index.ReservedFieldNameNamespace,
```
Comment (richardartoul) [inline storage/index.go]: Even if this namespace is really long, the index will end up compressing the re-used namespace pretty well right?

File: storage/index.go [hunk 9]
```diff
@@ -21,61 +21,260 @@
...
+		Fields: fields,
+	}, nil
 }
 
-func (n dbIndexNoOp) Query(context.Context, index.Query, index.QueryOptions) (index.QueryResults, error) {
-	return index.QueryResults{}, nil
+// NB(prateek): we take an independent copy of the bytes underlying
+// any ids provided, as we need to maintain the lifecycle of the indexed
+// bytes separately from the rest of the storage subsystem.
+func (i *dbIndex) clone(id ident.ID) []byte {
```
Comment (richardartoul) [inline storage/index.go]: Do you have to do this manual clone to prevent sharing pooled byte slices with the index?

File: storage/index_insert_queue.go [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+	sleepFn      func(time.Duration)
+	notifyInsert chan struct{}
+
+	metrics dbIndexInsertQueueMetrics
+}
+
+type newDatabaseIndexInsertQueueFn func(
+	dbIndexInsertBatchFn, clock.NowFn, tally.Scope) databaseIndexInsertQueue
+
+// FOLLOWUP(prateek): subsequent PR to wire up rate limiting to runtime.Options
```
Comment (richardartoul) [inline storage/index_insert_queue.go]: Lets sync on where you want to put this. I still need to add runtime configuration for LRU size, and I think we want to standardize KV watcher locations

File: storage/index_insert_queue.go [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+	}
+
+	q.state = dbIndexInsertQueueStateOpen
+	go q.insertLoop()
+	return nil
+}
+
+func (q *dbIndexInsertQueue) Stop() error {
+	q.Lock()
+	defer q.Unlock()
```
Comment (richardartoul) [inline storage/index_insert_queue.go]: Do you want to hold the lock for the whole duration here? Seems like you'd want to:

1. Grab the lock
2. Mark the insert queue as closed
3. Release the lock

That way callers that are trying to call Write() can receive an immediate error instead of waiting for a while on the lock only to be told the insert queue is closed

File: storage/database.go [hunk 12]
```diff
@@ -138,21 +138,15 @@ func NewDatabase(
...
 	logger := iopts.Logger()
 
-	index := databaseIndexNoOp
-	if opts.IndexingEnabled() {
-		index, err = newDatabaseIndex(opts)
-		if err != nil {
-			return nil, err
-		}
-	}
+	var idx databaseIndex
```
Comment (robskillington) [inline storage/database.go]: Hm this is just nil yeah?  Is it worth explicitly setting just the default nil value on the `db` struct?

File: storage/index.go [hunk 13]
```diff
@@ -21,61 +21,260 @@
...
+	seg, err := mem.New(1, opts.MemSegmentOptions())
+	if err != nil {
+		return nil, err
+	}
+
+	idx := &dbIndex{
+		insertMode: dbIndexInsertAsync,
+		active: dbIndexBlock{
+			segment:    seg,
+			expiryTime: maxTime, // FOLLOWUP(prateek): undo hard-coding to infinite retention
```
Comment (robskillington) [inline storage/index.go]: Now that we are going to make this per namespace, can we just use the namespace retention options here?

File: storage/index.go [hunk 14]
```diff
@@ -21,61 +21,260 @@
...
+			numErr++
+		} else {
+			insert.fns.OnIndexSuccess(i.active.expiryTime)
+		}
+		// NB: we need to release held resources so we un-conditionally execute the Finalize.
+		insert.fns.OnIndexFinalize()
+	}
+
+	if numErr != 0 {
+		i.metrics.AsyncInsertErrors.Inc(numErr)
```
Comment (robskillington) [inline storage/index.go]: Might it be worth adding a `OnIndexError(...)` callback too?

File: storage/index.go [hunk 15]
```diff
@@ -21,61 +21,260 @@
...
+	if err != nil {
+		fns.OnIndexFinalize()
+		return err
+	}
+
+	i.RLock()
+	if !i.isOpenWithRLock() {
+		i.RUnlock()
+		i.metrics.InsertAfterClose.Inc(1)
+		fns.OnIndexFinalize()
```
Comment (robskillington) [inline storage/index.go]: Hm why do we finalize on error here but not in write batch method?  Also do we need a write batch and and non batched one?  Does this ever get used?

File: storage/index.go [hunk 16]
```diff
@@ -21,61 +21,260 @@
...
+
+func (i *dbIndex) doc(ns, id ident.ID, tags ident.Tags) (doc.Document, error) {
+	fields := make([]doc.Field, 0, 1+len(tags))
+	fields = append(fields, doc.Field{
+		Name:      index.ReservedFieldNameNamespace,
+		Value:     i.clone(ns),
+		ValueType: doc.StringValueType,
+	})
+	for j := 0; j < len(tags); j++ {
+		t := tags[j]
```
Comment (robskillington) [inline storage/index.go]: nit: Any reason not to `for j, t := range tags { ... }`?

File: storage/index/iterator.go [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+
+func (i *idsIter) wrapBytes(bytes []byte) checked.Bytes {
+	return i.wrapperPool.Get(bytes)
+}
+
+func (i *idsIter) parseAndStore(d doc.Document) {
+	i.currentID = i.idPool.BinaryID(i.wrapBytes(d.ID))
+	nsFound := false
+	for _, f := range d.Fields {
+		if !nsFound && bytes.Equal(f.Name, ReservedFieldNameNamespace) {
```
Comment (robskillington) [inline storage/index/iterator.go]: Yeah if the fields were separated from normal fields then we wouldn't have to do this search, or at the very least be search a smaller bucket of tags that are definitely the reserved tags.

File: storage/index/iterator.go [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+			nsFound = true
+			continue
+		}
+		i.currentTags = append(i.currentTags,
+			i.idPool.BinaryTag(i.wrapBytes(f.Name), i.wrapBytes(f.Value)))
+	}
+
+	// should have found ns
+	if !nsFound {
+		i.err = errInvalidResultMissingNamespace
```
Comment (robskillington) [inline storage/index/iterator.go]: Shouldn't this return `false` in `Next()` if this happens?  Seems like `i.err` is not checked after calling `i.parseAndStore(....)`, perhaps we do that and add a test to catch this?

File: storage/index_insert_queue.go (hunk 11 above)
Comment (robskillington) [reply storage/index_insert_queue.go]: There's no blocking in the code below, there's is a best effort to kick off the final flush, so whether you release it slightly earlier there will be no real difference in how fast this method executes.

File: storage/index_insert_queue.go [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+	dbIndexInsertQueueStateOpen
+	dbIndexInsertQueueStateClosed
+)
+
+var (
+	defaultIndexBatchBackoff   = time.Second
+	defaultIndexPerSecondLimit = 10000
+)
+
+type dbIndexInsertQueue struct {
```
Comment (robskillington) [inline storage/index_insert_queue.go]: Could you possibly use Genny to condense needing a copy?  After looking at a lot of stuff, actually doesn't seem like much is different between the two at all.

File: storage/types.go [hunk 20]
```diff
@@ -411,6 +413,40 @@ type databaseIndex interface {
...
+	// inserts to the index asynchronously. It executes the provided callbacks
+	// based on the result of the execution. The returned wait group can be used
+	// if the insert is required to be synchronous.
+	Insert(d doc.Document, s indexInsertLifecycleHooks) (*sync.WaitGroup, error)
+}
+
+// indexInsertLifecycleHooks provides a set of callback hooks to allow the
+// databaseIndex to do lifecycle management of any resources retained
+// during indexing.
+type indexInsertLifecycleHooks interface {
```
Comment (robskillington) [inline storage/types.go]: Yeah we can we rename this something like `OnIndexSeries`?  This would be similar to `OnRetrieveBlock(...)` and `OnReadBlock` type, etc:
https://github.com/m3db/m3db/blob/352f2364f7202e591c69017b33cf78e6b0885f8d/storage/block/types.go#L219-L230

File: storage/index.go (hunk 1 above)
Comment (prateek) [reply storage/index.go]: It's the unique identifier, I'll rework to use created at nanos.

File: storage/index.go (hunk 4 above)
Comment (prateek) [reply storage/index.go]: sure thing, renamed to `newIndexQueueFn`

File: storage/index.go (hunk 6 above)
Comment (prateek) [reply storage/index.go]: Nah it will. I'll draw out the call chains for the different paths here:

- For new entry (previously unseen in the shard):
```
shard.WriteTagged()
  => shardInsertQueue.Insert()
  => shard.writeBatch()
  => index.Write()
  => indexQueue.Insert()
  => index.writeBatch() 
```

- For entry which exists in the shard, but needs indexing (either past the TTL or the last indexing hasn't happened/failed):
```
shard.WriteTagged()
  => index.Write()
  => indexQueue.Insert()
  => index.writeBatch() 
```

File: storage/index.go (hunk 7 above)
Comment (prateek) [reply storage/index.go]: The latter, it's a completely different lifecycle. Going to rely on the GC to clear as far as we can. Will change if needed down the road.

File: storage/index_insert_queue.go (hunk 10 above)
Comment (prateek) [reply storage/index_insert_queue.go]: Sure i'll bring it up in person tomorrow.

File: storage/index_insert_queue.go (hunk 11 above)
Comment (prateek) [reply storage/index_insert_queue.go]: The latency spike during the close path is such an edge case, don't need think about it while there's lots of core functionality to be developed.

File: storage/index.go (hunk 9 above)
Comment (prateek) [reply storage/index.go]: Kinda -t he storage code finalises IDs when it expires stuff, so we can't rely on the ident passed to the index to be valid for the entirety of the lifecycle of the index. The clone allows us to bypass that restriction. 

Ideally, we'd maintain only a single reference for the ident and do the ref counting correctly to only release once both the storage and index systems were done with it, but that's a much more pervasive change than I wanted to make to the storage side. Did this as a comprise.

File: storage/index.go (hunk 2 above)
Comment (prateek) [reply storage/index.go]: Sure, i'll make it an option for now. Will do the YAML config when making structs for it in a subsequent PR.

File: storage/index.go (hunk 1 above)
Comment (prateek) [reply storage/index.go]: done

File: storage/index.go (hunk 2 above)
Comment (prateek) [reply storage/index.go]: done

File: storage/index.go (hunk 3 above)
Comment (prateek) [reply storage/index.go]: The original intent was to point out to readers of the code that skipping `OnIndexSuccess` was deliberate. Don't want to loose that. If you think the comment doesn't help, I can rephrase it to be clearer.

Comment (prateek) [reply storage/index.go]: done

File: storage/index.go (hunk 5 above)
Comment (prateek) [reply storage/index.go]: Haven't done any benchmarking for it, but you can imagine worst case is it's in the order of 10s to 100s of milliseconds. I can rephrase the comment to be clearer. What about it was confusing?

File: storage/index.go (hunk 8 above)
Comment (prateek) [reply storage/index.go]: Yeah either implementation of segment will keep a bitset of documents per unique (fieldname, fieldvalue)

File: storage/database.go (hunk 12 above)
Comment (prateek) [reply storage/database.go]: Sure, moved to the namespace and made the change.

File: storage/index.go (hunk 13 above)
Comment (prateek) [reply storage/index.go]: sure, will take as a ctor parameter and pass down.

File: storage/index.go (hunk 14 above)
Comment (prateek) [reply storage/index.go]: I didn't do it because it doesn't enable anything special in the shard entry once you're notified there's an indexing error. We emit a metric for it in the index-insert-queue/index already. And we'd retry the indexing on the next write for the shard entry as is.

Is there something else you'd like to do in the shard entry on `OnIndexError`?

File: storage/index.go (hunk 15 above)
Comment (prateek) [reply storage/index.go]: At this point the write has not made it to the index queue. We've received the write and realised the index has been closed so we terminate early.

Comment (prateek) [reply storage/index.go]: Re: batch v non-batch. This thread explains the call chains: https://github.com/m3db/m3db/pull/514#discussion_r177185849 

Don't need batch v non-batch because all writes are batched internally by the queue.

File: storage/index.go (hunk 16 above)
Comment (prateek) [reply storage/index.go]: None, I'll make the change.

File: storage/index.go (hunk 13 above)
Comment (prateek) [reply storage/index.go]: done

File: storage/index.go (hunk 16 above)
Comment (prateek) [reply storage/index.go]: done

File: storage/index/iterator.go (hunk 17 above)
Comment (prateek) [reply storage/index/iterator.go]: For sure.

File: storage/index/iterator.go (hunk 18 above)
Comment (prateek) [reply storage/index/iterator.go]: Good catch. Addressed and added a test.

File: storage/index_insert_queue.go (hunk 19 above)
Comment (prateek) [reply storage/index_insert_queue.go]: Yeah they definitely share over 90% of code right now. I'm hesitant to make the change just now - I'd like to see how things are after we've got a fully functioning index. I'll open an issue to track for now, can come back and cleanup after.

Comment (prateek) [reply storage/index_insert_queue.go]: https://github.com/m3db/m3db/issues/517

File: storage/types.go (hunk 20 above)
Comment (prateek) [reply storage/types.go]: Sure thing.

File: storage/index.go (hunk 6 above)
Comment (richardartoul) [reply storage/index.go]: ok makes sense.

This was super helpful, can we stick that call chain somewhere

File: storage/index.go (hunk 8 above)
Comment (richardartoul) [reply storage/index.go]: Not sure I'm totally following. I guess my question was if you insert a million document, we won't have a million different ident.ID's referring to the same namespace right?

File: storage/index.go (hunk 9 above)
Comment (richardartoul) [reply storage/index.go]: Fair enough, I'll have to look back at your code to see what you are cloning, but I think the series.ID is never finalized and would probably be safe to share with the index (several parts of the codebase depend on series.ID not being manually finalized)

File: storage/index_insert_queue.go (hunk 11 above)
Comment (richardartoul) [reply storage/index_insert_queue.go]: fair enough

File: storage/index.go [hunk 21]
```diff
@@ -21,61 +21,255 @@
...
+	return i.insertQueue.Stop()
 }
 
-func (n dbIndexNoOp) Query(context.Context, index.Query, index.QueryOptions) (index.QueryResults, error) {
-	return index.QueryResults{}, nil
+func (i *nsIndex) doc(id ident.ID, tags ident.Tags) (doc.Document, error) {
+	fields := make([]doc.Field, 0, 1+len(tags))
+	fields = append(fields, doc.Field{
+		Name:      index.ReservedFieldNameID,
+		Value:     i.clone(id),
```
Comment (richardartoul) [inline storage/index.go]: This is the series ID? I think we can possibly avoid cloning this because it never gets finalized (WiredList and CommitLog and possibly some other things also depend on this behavior)

Comment (richardartoul) [reply storage/index.go]: @robskillington

File: storage/index_insert_queue.go [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+	"github.com/m3db/m3db/clock"
+	"github.com/m3db/m3ninx/doc"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	errIndexInsertQueueNotOpen             = errors.New("index insert queue is not open")
+	errIndexInsertQueueAlreadyOpenOrClosed = errors.New("index insert queue already open or is closed")
+	errNewSeriesIndexRateLimitExceeded     = errors.New("indexing new series eclipses rate limit")
```
Comment (richardartoul) [inline storage/index_insert_queue.go]: extremely nit: eclipses seems like an expensive word where exceeds would suffice

File: storage/index_insert_queue.go [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+
+var (
+	defaultIndexBatchBackoff   = time.Second
+	defaultIndexPerSecondLimit = 10000
+)
+
+type nsIndexInsertQueue struct {
+	sync.RWMutex
+	state nsIndexInsertQueueState
+	// rate limits
```
Comment (richardartoul) [inline storage/index_insert_queue.go]: super nit, if these have distinction "sections" a space would help along with the comment

File: storage/index.go (hunk 6 above)
Comment (prateek) [reply storage/index.go]: done

File: storage/index.go (hunk 9 above)
Comment (prateek) [reply storage/index.go]: I think you're right but I'd still prefer to explicitly copy. It makes the lifecycle a lot easier to reason about, and avoids issues if we ever do change to start finalizing ids.

File: storage/index_insert_queue.go (hunk 10 above)
Comment (prateek) [reply storage/index_insert_queue.go]: talked offline. we should do this when one of us has cycles.

File: storage/index.go (hunk 21 above)
Comment (prateek) [reply storage/index.go]: same as above:  I think you're right but I'd still prefer to explicitly copy. It makes the lifecycle a lot easier to reason about, and avoids issues if we ever do change to start finalizing ids.

Comment (richardartoul) [reply storage/index.go]: I guess if the index block size is different than m3db's blocksize that makes sense

File: storage/database.go [hunk 24]
```diff
@@ -104,16 +107,23 @@ type databaseMetrics struct {
...
+	indexDisabledScope := scope.SubScope("index-disabled")
 	return databaseMetrics{
 		unknownNamespaceRead:                unknownNamespaceScope.Counter("read"),
 		unknownNamespaceWrite:               unknownNamespaceScope.Counter("write"),
 		unknownNamespaceWriteTagged:         unknownNamespaceScope.Counter("write-tagged"),
 		unknownNamespaceFetchBlocks:         unknownNamespaceScope.Counter("fetch-blocks"),
 		unknownNamespaceFetchBlocksMetadata: unknownNamespaceScope.Counter("fetch-blocks-metadata"),
+		unknownNamespaceQueryIDs:            unknownNamespaceScope.Counter("query-ids"),
+		queryIDsIndexDisabled:               indexDisabledScope.Counter("query-ids"),
+		writeTaggedIndexDisabled:            indexDisabledScope.Counter("write-tagged"),
```
Comment (robskillington) [inline storage/database.go]: nit: rename these to be explicit about them being error cases?  `errQueryIDsIndexDisabled`, etc?  And for the names perhaps "err-query-ids" and "err-write-tagged"?

File: storage/index_insert_queue.go (hunk 22 above)
Comment (prateek) [reply storage/index_insert_queue.go]: lol it's a copy of the existing code from the shard insert queue. I'll change both.

File: storage/index_insert_queue.go (hunk 23 above)
Comment (prateek) [reply storage/index_insert_queue.go]: sure thing

File: storage/index.go (hunk 21 above)
Comment (robskillington) [reply storage/index.go]: Yeah copying should be ok, it's only on insert thankfully.

File: storage/index.go [hunk 25]
```diff
@@ -21,61 +21,272 @@
...
+}
+
+// NB(prateek): including the call chains leading to this point:
+//
+// - For new entry (previously unseen in the shard):
+//     shard.WriteTagged()
+//       => shardInsertQueue.Insert()
+//       => shard.writeBatch()
+//       => index.Write()
+//       => indexQueue.Insert()
```
Comment (robskillington) [inline storage/index.go]: Is there a reason we don't just `index.writeBatch` from `shard.writeBatch`?  Seems like a second level of queuing that may not be necessary?

File: storage/index.go (hunk 14 above)
Comment (robskillington) [reply storage/index.go]: Hm probably not at this time, just seemed like having a symmetrical API would've made it clear we don't do anything when we encounter an error on insert.

File: storage/index/options.go [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+	wrapperPool    xpool.CheckedBytesWrapperPool
+}
+
+// NewOptions returns a new index.Options object with default properties.
+func NewOptions() Options {
+	idPool := ident.NewPool(nil, nil)
+	wrapperPool := xpool.NewCheckedBytesWrapperPool(defaultCheckedBytesPoolInitialSize)
+	wrapperPool.Init()
+	return &opts{
+		insertMode:     InsertAsync,
```
Comment (robskillington) [inline storage/index/options.go]: Hm I believe that for defaults for series inserts it is synchronous, we should match the defaults of that with this - if writes are sync by default, so too should be indexing yeah?

Comment (robskillington) [reply storage/index/options.go]: Also can you make this a default const at the top of this file so it's more clear. i.e. `defaultInsertMode = XYZ`

File: storage/index_insert_queue.go [hunk 27]
```diff
@@ -0,0 +1,258 @@
...
+	nsIndexInsertQueueStateOpen
+	nsIndexInsertQueueStateClosed
+)
+
+var (
+	defaultIndexBatchBackoff   = time.Second
+	defaultIndexPerSecondLimit = 10000
+)
+
+type nsIndexInsertQueue struct {
```
Comment (robskillington) [inline storage/index_insert_queue.go]: What was the outcome on determining if `genny` could be used here or not?

File: storage/namespace.go [hunk 28]
```diff
@@ -495,6 +498,17 @@ func (n *dbNamespace) WriteTagged(
 	return err
 }
 
+func (n *dbNamespace) QueryIDs(
+	ctx context.Context,
+	query index.Query,
+	opts index.QueryOptions,
+) (index.QueryResults, error) {
+	callStart := n.nowFn()
+	res, err := n.reverseIndex.Query(ctx, query, opts)
```
Comment (robskillington) [inline storage/namespace.go]: Hm this will panic at this point as `reverseIndex` is nil, should we instead for now `return fmt.Errorf("not implemented")` or something?

File: storage/index.go (hunk 25 above)
Comment (prateek) [reply storage/index.go]: talked offline, going to keep the two level indexing going to allow for batch'd writes to the index.

File: storage/index/options.go (hunk 26 above)
Comment (prateek) [reply storage/index/options.go]: sure, will do

File: storage/index_insert_queue.go (hunk 27 above)
Comment (prateek) [reply storage/index_insert_queue.go]: Going to do it once the APIs stabilise a bit more. Have  https://github.com/m3db/m3db/issues/517 to track.

File: storage/namespace.go (hunk 28 above)
Comment (prateek) [reply storage/namespace.go]: This can't happen right now because the only caller to this method is `db.QueryIDs` which does the equivalent check before calling `ns.QueryIDs`.

Comment (prateek) [reply storage/namespace.go]: Added an error check and log statement to be defensive.

File: storage/index/iterator.go [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+
+func (i *idsIter) wrapBytes(bytes []byte) checked.Bytes {
+	return i.wrapperPool.Get(bytes)
+}
+
+func (i *idsIter) parseAndStore(d doc.Document) error {
+	idFound := false
+	for _, f := range d.Fields {
+		if !idFound && bytes.Equal(f.Name, ReservedFieldNameID) {
+			i.currentID = i.idPool.BinaryID(i.wrapBytes(f.Value))
```
Comment (robskillington) [inline storage/index/iterator.go]: aside: This whole wrap bytes thing makes me think maybe we should've been using `checked.Bytes` in the m3ninx code, but that's neither here nor there now.

File: storage/index/types.go [hunk 30]
```diff
@@ -23,8 +23,27 @@ package index
...
 	"github.com/m3db/m3ninx/index/segment"
+	"github.com/m3db/m3ninx/index/segment/mem"
 	"github.com/m3db/m3x/ident"
+	"github.com/m3db/m3x/instrument"
+)
+
+var (
+	// ReservedFieldNameID is the field name used to index the ID in the
+	// m3ninx subsytem.
+	ReservedFieldNameID = []byte("_m3db-id")
```
Comment (robskillington) [inline storage/index/types.go]: nit: Can we use the same special character here, i.e. `_m3db_id`?  Just when debugging things typing "_m3db_id" over and over again is much easier than "_m3db-id" (I've noticed this when debugging strange JSON structures, thank you human brain...).

Comment (prateek) [reply storage/index/types.go]: haha sure thing

File: storage/index/iterator.go (hunk 29 above)
Comment (prateek) [reply storage/index/iterator.go]: Not so sure about that. If we're going to have external consumers of `m3ninx`, the `[]byte` interface is a lot more familiar to Go users than `checked.Bytes` would be. 

Can chat more offline.
